File: php/tests/Metadata/EntityMapLoaderTest.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Metadata;

use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use SimpleOrm\Metadata\EntityIndex;
use SimpleOrm\Metadata\EntityMap;
use SimpleOrm\Metadata\EntityMapBuilder;
use SimpleOrm\Metadata\EntityMapLoader;
use SimpleOrm\Metadata\KeyStrategy;
use SimpleOrm\Metadata\MappingOptions;
use SimpleOrm\Metadata\PropertyMap;
use SimpleOrm\Metadata\RelationKind;
use SimpleOrm\Metadata\RelationshipKind;
use SimpleOrm\Metadata\RelationshipMap;
use SimpleOrm\Tests\Sample\Models\DailySales;
use SimpleOrm\Tests\Sample\Models\MonthlySalesTotal;
use SimpleOrm\Tests\Sample\Models\Role;
use SimpleOrm\Tests\Sample\Models\Transaction;
use SimpleOrm\Tests\Sample\Models\TransactionDetail;
use SimpleOrm\Tests\Sample\Models\User;
use SimpleOrm\Tests\Sample\Models\UserActivityReport;
use SimpleOrm\Tests\Sample\Models\UserProfile;
use SimpleOrm\Tests\Sample\Models\UserRole;
use SimpleOrm\Tests\Sample\Models\UserTransactionTotal;

/**
 * Happy-path metadata loading across the sample models: loader precedence
 * (explicit → attribute → convention), caching, key strategies, and the
 * inherited-property ordering rule (spec/metadata-model.md "Loader
 * precedence": most-derived class first, base classes' columns last).
 */
final class EntityMapLoaderTest extends TestCase
{
    private EntityMapLoader $loader;

    protected function setUp(): void
    {
        $this->loader = new EntityMapLoader();
    }

    #[Test]
    public function user_maps_table_generated_key_and_inherited_audit_columns(): void
    {
        $map = $this->loader->load(User::class);

        self::assertSame(RelationKind::Table, $map->kind);
        self::assertSame('users', $map->relationName);
        self::assertSame(KeyStrategy::DatabaseGenerated, $map->keyStrategy);
        self::assertSame(['id'], array_map(self::columnName(...), $map->keyProperties));

        // Derived class columns first, BaseModel audit columns last.
        self::assertSame(
            ['id', 'name', 'email', 'display_name', 'created_at', 'updated_at'],
            array_map(self::columnName(...), $map->properties),
        );
        self::assertFalse($map->property('name')?->nullable);
        self::assertTrue($map->property('updatedAtUtc')?->nullable);

        self::assertCount(2, $map->indexes);
        self::assertTrue(self::indexNamed($map, 'ix_users_email')->unique);
        self::assertFalse(self::indexNamed($map, 'ix_users_display_name')->unique);
    }

    #[Test]
    public function user_role_maps_composite_natural_key_and_two_relationships(): void
    {
        $map = $this->loader->load(UserRole::class);

        self::assertSame(KeyStrategy::Natural, $map->keyStrategy);
        self::assertSame(['user_id', 'role_id'], array_map(self::columnName(...), $map->keyProperties));
        self::assertCount(2, $map->relationships);

        $user = self::relationshipNamed($map, 'user');
        self::assertSame(User::class, $user->targetType);
        $role = self::relationshipNamed($map, 'role');
        self::assertSame(['roleId'], $role->foreignKeyProperties);
    }

    #[Test]
    public function transaction_maps_version_indexes_and_foreign_key(): void
    {
        $map = $this->loader->load(Transaction::class);

        self::assertSame('version', $map->versionProperty?->columnName);
        self::assertSame(User::class, $map->property('userId')?->foreignKeyReferences);

        self::assertCount(2, $map->indexes);
        self::assertSame('ix_transactions_user_id', $map->indexes[0]->name);
        $named = $map->indexes[1];
        self::assertSame('ix_transactions_status_created', $named->name);
        self::assertSame([false, true], array_map(static fn ($c) => $c->descending, $named->columns));

        $manyToOne = self::relationshipNamed($map, 'user');
        self::assertSame(RelationshipKind::ManyToOne, $manyToOne->kind);
        self::assertSame(['userId'], $manyToOne->foreignKeyProperties);

        $details = self::relationshipNamed($map, 'details');
        self::assertSame(RelationshipKind::OneToMany, $details->kind);
        self::assertSame(TransactionDetail::class, $details->targetType);
    }

    #[Test]
    public function view_and_materialized_view_map_with_their_capabilities(): void
    {
        $view = $this->loader->load(UserTransactionTotal::class);
        self::assertSame(RelationKind::View, $view->kind);
        self::assertSame(KeyStrategy::Natural, $view->keyStrategy);
        self::assertSame([], $view->indexes);

        $materialized = $this->loader->load(MonthlySalesTotal::class);
        self::assertSame(RelationKind::MaterializedView, $materialized->kind);
        self::assertCount(1, $materialized->indexes);
        self::assertTrue($materialized->indexes[0]->unique);
    }

    #[Test]
    public function statement_maps_sql_and_declared_parameters(): void
    {
        $map = $this->loader->load(DailySales::class);

        self::assertSame(RelationKind::Statement, $map->kind);
        self::assertNull($map->relationName);
        self::assertStringContainsString('@since', (string) $map->definingSql);
        self::assertSame(KeyStrategy::None, $map->keyStrategy);

        self::assertCount(1, $map->statementParameters);
        self::assertSame('since', $map->statementParameters[0]->name);
    }

    #[Test]
    public function procedure_maps_keyless_and_nullable_columns(): void
    {
        $map = $this->loader->load(UserActivityReport::class);

        self::assertSame(RelationKind::Procedure, $map->kind);
        self::assertSame('user_activity_report', $map->relationName);
        self::assertSame(KeyStrategy::None, $map->keyStrategy);
        self::assertStringContainsString('@since', (string) $map->definingSql);
        self::assertSame('since', $map->statementParameters[0]->name);
        self::assertTrue($map->property('lastTransactionAtUtc')?->nullable);
    }

    #[Test]
    public function role_and_user_profile_declare_the_remaining_navigation_kinds(): void
    {
        $role = $this->loader->load(Role::class);
        $manyToMany = self::relationshipNamed($role, 'users');
        self::assertSame(RelationshipKind::ManyToMany, $manyToMany->kind);
        self::assertSame(User::class, $manyToMany->targetType);
        self::assertSame(UserRole::class, $manyToMany->linkType);
        self::assertSame(['roleId'], $manyToMany->linkForeignKeysToOwner);
        self::assertSame(['userId'], $manyToMany->linkForeignKeysToTarget);

        $user = $this->loader->load(User::class);
        $profile = self::relationshipNamed($user, 'profile');
        self::assertSame(RelationshipKind::OneToOne, $profile->kind);
        self::assertSame(UserProfile::class, $profile->targetType);
        self::assertSame(['userId'], $profile->foreignKeyProperties);
    }

    #[Test]
    public function maps_are_cached_per_loader(): void
    {
        self::assertSame($this->loader->load(User::class), $this->loader->load(User::class));
    }

    #[Test]
    public function convention_loader_maps_an_unannotated_type(): void
    {
        $map = $this->loader->load(Person::class);

        self::assertSame(RelationKind::Table, $map->kind);
        self::assertSame('person', $map->relationName);
        self::assertSame(KeyStrategy::DatabaseGenerated, $map->keyStrategy);
        self::assertSame(['id', 'first_name'], array_map(self::columnName(...), $map->properties));
        self::assertTrue($map->property('firstName')?->nullable);
    }

    #[Test]
    public function explicit_registration_wins_over_attributes(): void
    {
        $builder = EntityMapBuilder::for(User::class)->table('users_manual');
        $builder->column('id', key: true, generated: true);
        $builder->column('name');

        $options = new MappingOptions(explicitMaps: [User::class => $builder->build()]);
        $map = (new EntityMapLoader($options))->load(User::class);

        self::assertSame('users_manual', $map->relationName);
        self::assertCount(2, $map->properties);
    }

    private static function columnName(PropertyMap $property): string
    {
        return $property->columnName;
    }

    private static function indexNamed(EntityMap $map, string $name): EntityIndex
    {
        foreach ($map->indexes as $index) {
            if ($index->name === $name) {
                return $index;
            }
        }

        self::fail("no index named '{$name}'");
    }

    private static function relationshipNamed(
        EntityMap $map,
        string $propertyName,
    ): RelationshipMap {
        foreach ($map->relationships as $relationship) {
            if ($relationship->propertyName === $propertyName) {
                return $relationship;
            }
        }

        self::fail("no relationship for property '{$propertyName}'");
    }
}

// Deliberately no SimpleOrm attributes: the convention-loader fixture.
final class Person
{
    public int $id;

    public ?string $firstName = null;
}

File: go/orm/internal/metadata/entity_map_loader_test.go
package metadata_test

import (
	"reflect"
	"testing"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/metadata"
	"github.com/kosmassp/SimpleOrm/go/orm/sample"
)

// Happy-path metadata loading across the sample models (mirrors
// dotnet/tests/SimpleOrm.Tests/EntityMapLoaderTests.cs).

func columnNames(m *core.EntityMap) []string {
	names := make([]string, len(m.Properties))
	for i, p := range m.Properties {
		names[i] = p.ColumnName
	}
	return names
}

func equalStrings(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func propertyByColumn(m *core.EntityMap, column string) *core.PropertyMap {
	for _, p := range m.Properties {
		if p.ColumnName == column {
			return p
		}
	}
	return nil
}

func TestUser_MapsTableGeneratedKeyAndInheritedAuditColumns(t *testing.T) {
	loader := metadata.NewLoader(nil)
	m, err := metadata.Load[sample.User](loader)
	if err != nil {
		t.Fatal(err)
	}

	if m.Kind != core.RelationTable {
		t.Errorf("Kind = %v, want Table", m.Kind)
	}
	if m.RelationName != "users" {
		t.Errorf("RelationName = %q, want users", m.RelationName)
	}
	if m.KeyStrategy != core.KeyDatabaseGenerated {
		t.Errorf("KeyStrategy = %v, want DatabaseGenerated", m.KeyStrategy)
	}
	if len(m.KeyProperties) != 1 || m.KeyProperties[0].ColumnName != "id" {
		t.Errorf("KeyProperties = %+v", m.KeyProperties)
	}

	want := []string{"id", "name", "email", "display_name", "created_at", "updated_at"}
	if got := columnNames(m); !equalStrings(got, want) {
		t.Errorf("columns = %v, want %v (own fields first, base fields last)", got, want)
	}

	if propertyByColumn(m, "name").IsNullable {
		t.Error("name must not be nullable")
	}
	if !propertyByColumn(m, "updated_at").IsNullable {
		t.Error("updated_at must be nullable")
	}

	if len(m.Indexes) != 2 {
		t.Fatalf("Indexes = %+v, want 2", m.Indexes)
	}
	for _, idx := range m.Indexes {
		switch idx.Name {
		case "ix_users_email":
			if !idx.Unique {
				t.Error("ix_users_email must be unique")
			}
		case "ix_users_display_name":
			if idx.Unique {
				t.Error("ix_users_display_name must not be unique")
			}
		default:
			t.Errorf("unexpected index %q", idx.Name)
		}
	}
}

func TestUserRole_MapsCompositeNaturalKeyAndTwoRelationships(t *testing.T) {
	loader := metadata.NewLoader(nil)
	m, err := metadata.Load[sample.UserRole](loader)
	if err != nil {
		t.Fatal(err)
	}

	if m.KeyStrategy != core.KeyNatural {
		t.Errorf("KeyStrategy = %v, want Natural", m.KeyStrategy)
	}
	keyColumns := make([]string, len(m.KeyProperties))
	for i, k := range m.KeyProperties {
		keyColumns[i] = k.ColumnName
	}
	if !equalStrings(keyColumns, []string{"user_id", "role_id"}) {
		t.Errorf("key columns = %v", keyColumns)
	}
	if len(m.Relationships) != 2 {
		t.Fatalf("Relationships = %+v, want 2", m.Relationships)
	}
	for _, r := range m.Relationships {
		if r.PropertyName == "User" && r.TargetType != reflect.TypeFor[sample.User]() {
			t.Errorf("User relationship target = %v", r.TargetType)
		}
		if r.PropertyName == "Role" && !equalStrings(r.ForeignKeyProperties, []string{"RoleID"}) {
			t.Errorf("Role relationship FK = %v", r.ForeignKeyProperties)
		}
	}
}

func TestTransaction_MapsVersionIndexesAndForeignKey(t *testing.T) {
	loader := metadata.NewLoader(nil)
	m, err := metadata.Load[sample.Transaction](loader)
	if err != nil {
		t.Fatal(err)
	}

	if m.VersionProperty == nil || m.VersionProperty.ColumnName != "version" {
		t.Errorf("VersionProperty = %+v", m.VersionProperty)
	}
	if propertyByColumn(m, "user_id").ForeignKeyReferences != reflect.TypeFor[sample.User]() {
		t.Errorf("user_id.ForeignKeyReferences = %v", propertyByColumn(m, "user_id").ForeignKeyReferences)
	}

	if len(m.Indexes) != 2 {
		t.Fatalf("Indexes = %+v, want 2", m.Indexes)
	}
	var named *core.EntityIndex
	for _, idx := range m.Indexes {
		if idx.Name == "ix_transactions_status_created" {
			named = idx
		}
	}
	if named == nil {
		t.Fatal("ix_transactions_status_created not found")
	}
	if len(named.Columns) != 2 || named.Columns[0].Descending || !named.Columns[1].Descending {
		t.Errorf("named index columns = %+v", named.Columns)
	}

	var manyToOne *core.RelationshipMap
	for _, r := range m.Relationships {
		if r.Kind == core.RelationshipManyToOne {
			manyToOne = r
		}
	}
	if manyToOne == nil || !equalStrings(manyToOne.ForeignKeyProperties, []string{"UserID"}) {
		t.Errorf("many-to-one FK = %+v", manyToOne)
	}
}

func TestViewAndMaterializedView_MapWithTheirCapabilities(t *testing.T) {
	loader := metadata.NewLoader(nil)

	view, err := metadata.Load[sample.UserTransactionTotal](loader)
	if err != nil {
		t.Fatal(err)
	}
	if view.Kind != core.RelationView {
		t.Errorf("Kind = %v, want View", view.Kind)
	}
	if view.KeyStrategy != core.KeyNatural {
		t.Errorf("KeyStrategy = %v, want Natural", view.KeyStrategy)
	}
	if len(view.Indexes) != 0 {
		t.Errorf("view indexes = %+v, want none", view.Indexes)
	}

	materialized, err := metadata.Load[sample.MonthlySalesTotal](loader)
	if err != nil {
		t.Fatal(err)
	}
	if materialized.Kind != core.RelationMaterializedView {
		t.Errorf("Kind = %v, want MaterializedView", materialized.Kind)
	}
	if len(materialized.Indexes) != 1 || !materialized.Indexes[0].Unique {
		t.Errorf("materialized view indexes = %+v", materialized.Indexes)
	}
}

func TestStatement_MapsSQLAndDeclaredParameters(t *testing.T) {
	loader := metadata.NewLoader(nil)
	m, err := metadata.Load[sample.DailySales](loader)
	if err != nil {
		t.Fatal(err)
	}

	if m.Kind != core.RelationStatement {
		t.Errorf("Kind = %v, want Statement", m.Kind)
	}
	if m.RelationName != "" {
		t.Errorf("RelationName = %q, want empty", m.RelationName)
	}
	if m.KeyStrategy != core.KeyNone {
		t.Errorf("KeyStrategy = %v, want None", m.KeyStrategy)
	}
	if len(m.StatementParameters) != 1 || m.StatementParameters[0].Name != "since" {
		t.Fatalf("StatementParameters = %+v", m.StatementParameters)
	}
	if m.StatementParameters[0].Type.Name() != "Time" {
		t.Errorf("parameter type = %v, want time.Time", m.StatementParameters[0].Type)
	}
}

func TestProcedure_MapsKeylessAndNullableColumns(t *testing.T) {
	loader := metadata.NewLoader(nil)
	m, err := metadata.Load[sample.UserActivityReport](loader)
	if err != nil {
		t.Fatal(err)
	}

	if m.Kind != core.RelationProcedure {
		t.Errorf("Kind = %v, want Procedure", m.Kind)
	}
	if m.RelationName != "user_activity_report" {
		t.Errorf("RelationName = %q", m.RelationName)
	}
	if m.KeyStrategy != core.KeyNone {
		t.Errorf("KeyStrategy = %v, want None", m.KeyStrategy)
	}
	if len(m.StatementParameters) != 1 || m.StatementParameters[0].Name != "since" {
		t.Fatalf("StatementParameters = %+v", m.StatementParameters)
	}
	if !propertyByColumn(m, "last_transaction_at_utc").IsNullable {
		t.Error("last_transaction_at_utc must be nullable")
	}
}

func TestLoader_MapsAreCachedPerLoader(t *testing.T) {
	loader := metadata.NewLoader(nil)
	a, err := metadata.Load[sample.User](loader)
	if err != nil {
		t.Fatal(err)
	}
	b, err := metadata.Load[sample.User](loader)
	if err != nil {
		t.Fatal(err)
	}
	if a != b {
		t.Error("Load must return the same cached map for the same loader")
	}
}
